pkg/cluster/cluster.go

```diff
@@ -1108,3 +1114,49 @@ func (c *Controller) waitForHealthCheckAfterCreate(ctx context.Context, cluster
 	}
 	return nil
 }
+
+// maybeFixKubeConfigInsideContainer modifies the kubeconfig to set the KIND
+// apiserver's name and location to the name advertised on the `kind` docker
+// network (usually "$name-control-plane" where $name has the `kind-` prefix
+// removed), port 6443. This happens when the following conditions are true:
+//
+//   - The cluster to create is a KIND cluster
+//   - The effective DOCKER_HOST is `/var/run/docker.sock`
+//   - The environment variable $CTLPTL_IN_CONTAINER is set to a truthy value
+func (c *Controller) maybeFixKubeConfigInsideContainer(ctx context.Context, cluster *api.Cluster) error {
+	if cluster.Product != string(clusterid.ProductKIND) {
+		return nil
+	}
+
+	if c.dockerClient.DaemonHost() != "unix:///var/run/docker.sock" {
+		return nil
+	}
+
+	if inContainer, err := strconv.ParseBool(
+		strings.TrimSpace(os.Getenv("CTLPTL_IN_CONTAINER")),
+	); err != nil || !inContainer {
+		return nil
+	}
+
+	hostName, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("could not determine container host name: %w", err)
+	}
+
+	err = c.dockerClient.NetworkConnect(ctx, kindNetworkName(), hostName, nil)
```

Comment:
can we make this a method on the `Admin` interface? maybe 

```
func (a *Admin) DockerNetworkName()
```

there's nothing here that really seems specific to `kind`, seems like this approach would work on minikube and k3d too (though i think it's ok to punt on minikube and k3d for now)